### ISSUE TITLE
convert lora .bin weights to ckpt or safetensors

### ISSUE THREAD

**Issue by harrywang** [haofanwang/Easy-Lora-Handbook#1]
Another question: `convert_lora_safetensor_to_diffusers.py` converts safetensors to diffusers format. After I trained LoRA model, I have the following in the output folder and checkpoint subfolder:
<img width="238" alt="Screenshot 2023-02-20 at 7 07 51 PM" src="https://user-images.githubusercontent.com/595772/220215722-30edd279-a7d7-4574-b0ae-e4995f07973d.png">
<img width="230" alt="Screenshot 2023-02-20 at 7 07 33 PM" src="https://user-images.githubusercontent.com/595772/220215727-44b9e27f-374e-41db-bc70-d2eade3e8620.png">

How to convert them into safetensors like the ones I downloaded from civitai or huggingface so that I can use this via Automatica1111?

Thanks a lot!!

**Comment by harrywang** [comment]
https://github.com/huggingface/diffusers/blob/main/scripts/convert_diffusers_to_original_stable_diffusion.py I found this but have not tried yet. Will report back once I finish the testing.

**Comment by harrywang** [comment]
Thanks! I will try and report back.

**Comment by haofanwang** [comment]
@harrywang I guess I have understood your need, please check [this response](https://github.com/haofanwang/Lora-for-Diffusers/issues/2#issuecomment-1441268566).

**Comment by happypuppy20** [comment]
Hi @haofanwang , I'm having the SAME issue as @harrywang .
Basically, we trained using lora and got those 6 files, .bin file and a few others.

To use on Automatic1111 it seems like we need to have a different extension. reading your above response as well wasn't clearly gave me any insight how I can do it.

Hope a solution would be found :) Thanks!

**Comment by haofanwang** [comment]
Well, the main goal of this project is for developers using diffusers, not Automatic1111. But I try my best to help.

I'm not familiar with API of Automatic1111. But to be honest, only pytorch_lora_weights.bin saves the weights of LoRA. Other files in the folders are saved by accelerate as [here](https://github.com/huggingface/diffusers/blob/589faa8c881de32932faebf92d2d8007135294d6/examples/text_to_image/train_text_to_image_lora.py#L770). What is the exact file needed for diffusers? The folder or the single .bin file?

@happypuppy20

**Comment by happypuppy20** [comment]
True, I'm referring to Automatic1111 as it's easy for us to run many tests easily and fast.
Regards devs using diffusers, maybe you can tell me cause I wasn't able to find it anywhere else, but how I can inference the model with settings used by Automatic1111 like : 

prompts that include large Positive , large Negative , Sampler , CFG #, sampling #, Size 512x512

I couldn't find for the Negative or Sampler , CFG...

**Comment by haofanwang** [comment]
Please dive into diffusers for detailed usage. There you can find how to add negative prompt in inference.

**Comment by harrywang** [comment]
> **Update:**
> 
> Hi all, I just added a [new script](https://github.com/haofanwang/Lora-for-Diffusers/blob/main/format_convert.py) to support converting between .bin and .safetensors. You guys can check it out.
> 
> But as I have mentioned [here](https://github.com/haofanwang/Lora-for-Diffusers/issues/2#issuecomment-1442116903), it has many limitations and may not work as expected. Anyway, you can try it first. If not work, we suggest to save the full model to safetensors directly, instead of only saving LoRA weights in .bin.

Sorry for the slow response - I was traveling in the last few days. Thanks for the script - I tried and the conversion worked but when I load it into Automatic1111 the lora has no effect - I am not sure what went wrong.

**Comment by jndietz** [comment]
I'm not really familiar with the structure of diffusers or the ckpt/safetensors format.  But is it just a matter of mapping keys from one object to another?  I think @haofanwang has something we can start with.

I'll close this issue out because I finally understand a distinct difference between the formats expected by various apps/libraries.  I for some reason though they should all work together but diffusers has it's own ecosystem, and the Stable Diffusion webuis have their own expected format.  Though, confusingly, automatic1111 webui can use .bin from the textual inversion pipelines.

Anywho, until automatic1111 adds support for reading diffusers, I think the right thing to do now is add an example script -- maybe @haofanwang script as it get ironed out.  I agree it shouldn't be part of the core functionality of diffusers.

**Comment by jiayutom** [comment]
> > **Update:**
> > Hi all, I just added a [new script](https://github.com/haofanwang/Lora-for-Diffusers/blob/main/format_convert.py) to support converting between .bin and .safetensors. You guys can check it out.
> > But as I have mentioned [here](https://github.com/haofanwang/Lora-for-Diffusers/issues/2#issuecomment-1442116903), it has many limitations and may not work as expected. Anyway, you can try it first. If not work, we suggest to save the full model to safetensors directly, instead of only saving LoRA weights in .bin.
> 
> Sorry for the slow response - I was traveling in the last few days. Thanks for the script - I tried and the conversion worked but when I load it into Automatic1111 the lora has no effect - I am not sure what went wrong.

![1677651843392](https://user-images.githubusercontent.com/122880878/222061109-7d7f70df-288d-486e-a81d-26b3d2ed8476.png)
Hi, can you help me with this problem. I convert my lora weight with your script to safetensor, but i can not use the converted lora in webui. it show this error. can you tell me what went wrong here

**Comment by harrywang** [comment]
> > > **Update:**
> > > Hi all, I just added a [new script](https://github.com/haofanwang/Lora-for-Diffusers/blob/main/format_convert.py) to support converting between .bin and .safetensors. You guys can check it out.
> > > But as I have mentioned [here](https://github.com/haofanwang/Lora-for-Diffusers/issues/2#issuecomment-1442116903), it has many limitations and may not work as expected. Anyway, you can try it first. If not work, we suggest to save the full model to safetensors directly, instead of only saving LoRA weights in .bin.
> > 
> > 
> > Sorry for the slow response - I was traveling in the last few days. Thanks for the script - I tried and the conversion worked but when I load it into Automatic1111 the lora has no effect - I am not sure what went wrong.
> 
> ![1677652044281](https://user-images.githubusercontent.com/122880878/222061653-71971d40-bee5-47e6-b71d-385529443ea0.png)
> 
> Hi, can you help me with this problem. I convert my lora weight with your script to safetensor, but i can not use the converted lora in webui. it show this error. can you tell me what went wrong here

you can also try my script at https://github.com/huggingface/diffusers/issues/2326#issuecomment-1447341775, which works for me.

**Comment by StanislawKarnacky** [comment]
@harrywang, @haofanwang sorry for bothering you, but is there any chance to convert .bin to .safetensors LoRA trained on SDXL Dreambooth? Tried 5 scripts at least, keys mismatch in all cases. Trained with: [https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/train_dreambooth_lora_sdxl.py](url)

**Comment by alivemachine** [comment]
Have you ever figured out how to convert your .bin to a .safetensors lora? @StanislawKarnacky
